### PR TITLE
HOTFIX: poll with zero millis during restoration

### PR DESCRIPTION
**Pull request by guozhangwang**
Mirror of #4096 for 0.11.01.

During the restoration phase, when thread state is still in PARTITION_ASSIGNED not RUNNING yet, call poll() on the normal consumer with 0 millisecond timeout, to unblock the restoration of other tasks as soon as possible.

### CODE REVIEW

**Comment by guozhangwang** [comment]
@xvrl @dguy @mjsax @bbejeck

**Comment by bbejeck** [comment]
Thanks, @guozhangwang. Overall LGTM.  Can we add a test for this condition? (but I don't think it should hold up the PR)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (xvrl) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: since the restore loop will now be very tight, we should also change the log level in  `StoreChangelogReader.completed()` to `trace` to avoid unnecessary logging at the debug level

Comment (xvrl) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: we already changed it in trunk and 1.0 [here](https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java#L197) so I don't think this should be too controversial

**Comment by guozhangwang** [comment]
retest this please

**Comment by dguy** [comment]
retest this please

**Comment by dguy** [comment]
Thanks @guozhangwang, LGTM assuming we change the logging in `StoreChangeLogReader` as @xvrl has suggested.
Should we have a PR against trunk for this, too?

**Comment by ijuma** [comment]
@dguy Is it not https://github.com/apache/kafka/pull/4086 ?

**Comment by dguy** [comment]
Thanks @ijuma, didn't see it.

**Comment by dguy** [comment]
I think the test failures could be related to this PR

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: extract lines 519 - 534 into a method and call from here and line 498 above?

**Comment by bbejeck** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:120]: old debugging code?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordQueue.java:86]: as above.

**Comment by mjsax** [comment]
Did we consider to pause all partitions until all stores are recreated? We would still call `poll(0)` for liveness check, but `poll(0)` would not return any data and thus we would not process any data speeding up restore even further.

**Comment by xvrl** [comment]
@mjsax I believe that is the plan for 1.1 – @guozhangwang had already started something along those lines in #4096, but for 0.11.0.2 it's probably less risky to just set poll.ms to 0 until restore is done.

**Comment by mjsax** [comment]
Makes sense for `0.11.0.2` -- thanks for pointing out @xvrl

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
Jenkins failures are not relevant, merging to 0.11.0 now.

**Review by dguy — APPROVED** [review]
LGTM

**Comment by pulyaevskiy** [comment]
Any plans to get this released as 0.11.0.2 soon?

**Comment by mjsax** [comment]
There are currently no concrete plans to release a 0.11.0.2 version -- but 1.0.0 will be out shortly (note: Kafka Streams is backward compatible to older brokers to you can just update your app an run against older brokers if you wish).

If course, you can always start a discussion about a 0.11.0.2 release at the dev mailing list.

**Comment by pulyaevskiy** [comment]
@mjsax Thanks, good to know!

We'll probably downgrade to 0.10.x and wait for 1.0.0. 
Is there a tentative date(range) for 1.0.0?

**Comment by mjsax** [comment]
We did have multiple RC already -- if we stop finding bugs, it's just a matter of days...

**Comment by mjsax** [comment]
@pulyaevskiy `1.0.0` was release, as there will be also a `0.11.0.2` soon -- was announced today. First RC should be cut end of this week.

**Comment by pulyaevskiy** [comment]
Thanks for the update @mjsax ! Will check out 1.0.0 and see if it fixes our problem.

**Pull request closed without merging**
